From c90d74cefc1ef066bdd0207463f759de91837ebf Mon Sep 17 00:00:00 2001
From: Valentin Caron <valentin.caron@foss.st.com>
Date: Mon, 2 Sep 2024 15:14:21 +0200
Subject: [PATCH 1227/1695] plat-stm32mp2: scmi_server: implement clock
 protocol

scmi_server_scpfw can now retrieve clocks description from DT.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 0881a530808dedb07eb885ca1e34bc7d352023d1 plat-stm32mp2: scmi_server: implement clock protocol]
Signed-off-by: Valentin Caron <valentin.caron@foss.st.com>
Change-Id: I1d07caa2d3c46f5adf2f7a1a3731fae465e12285
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/401861
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/arch/arm/plat-stm32mp2/conf.mk           |   2 +
 .../arm/plat-stm32mp2/scmi_server_scpfw.c     |   8 +
 .../scmi-server/include/scmi_clock_consumer.h |  35 +++
 core/lib/scmi-server/scmi_clock_consumer.c    | 278 ++++++++++++++++++
 core/lib/scmi-server/sub.mk                   |   1 +
 5 files changed, 324 insertions(+)
 create mode 100644 core/lib/scmi-server/include/scmi_clock_consumer.h
 create mode 100644 core/lib/scmi-server/scmi_clock_consumer.c

diff --git a/core/arch/arm/plat-stm32mp2/conf.mk b/core/arch/arm/plat-stm32mp2/conf.mk
index 6d31971ca..5675a1056 100644
--- a/core/arch/arm/plat-stm32mp2/conf.mk
+++ b/core/arch/arm/plat-stm32mp2/conf.mk
@@ -90,6 +90,7 @@ $(call force,CFG_SCMI_CORTEXM_AGENT,n)
 $(call force,CFG_SCMI_MSG_DRIVERS,n)
 $(call force,CFG_SCMI_PTA,n)
 $(call force,CFG_SCMI_SCPFW,n)
+$(call force,CFG_SCMI_SERVER_CLOCK_CONSUMER,n)
 $(call force,CFG_SCMI_SERVER_PD_CONSUMER,n)
 $(call force,CFG_SCMI_SERVER_REGULATOR_CONSUMER,n)
 $(call force,CFG_STM32_BSEC_PTA,n)
@@ -192,6 +193,7 @@ CFG_WITH_TUI ?= y
 CFG_SCMI_MSG_DRIVERS ?= n
 CFG_SCMI_PTA ?= y
 CFG_SCMI_SCPFW ?= y
+CFG_SCMI_SERVER_CLOCK_CONSUMER ?= y
 CFG_SCMI_SERVER_PD_CONSUMER ?= y
 CFG_SCMI_SERVER_REGULATOR_CONSUMER ?= y
 
diff --git a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
index 442a9e73d..5a898971c 100644
--- a/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
+++ b/core/arch/arm/plat-stm32mp2/scmi_server_scpfw.c
@@ -15,6 +15,7 @@
 #include <libfdt.h>
 #include <mm/core_memprot.h>
 #include <mm/core_mmu.h>
+#include <scmi_clock_consumer.h>
 #include <scmi_pd_consumer.h>
 #include <scmi/scmi_server.h>
 #include <tee_api_types.h>
@@ -412,6 +413,13 @@ static TEE_Result optee_scmi_server_probe(const void *fdt, int parent_node,
 					panic("Error during power domains init"
 					      );
 				break;
+			case SCMI_PROTOCOL_ID_CLOCK:
+				res = optee_scmi_server_init_clocks(fdt,
+					protocol_ctx->dt_node, agent_cfg,
+					channel_cfg);
+				if (res)
+					panic("Error during clocks init");
+				break;
 			default:
 				EMSG("%s Unknown protocol ID: %#x",
 				     protocol_ctx->dt_name,
diff --git a/core/lib/scmi-server/include/scmi_clock_consumer.h b/core/lib/scmi-server/include/scmi_clock_consumer.h
new file mode 100644
index 000000000..c27cfceb3
--- /dev/null
+++ b/core/lib/scmi-server/include/scmi_clock_consumer.h
@@ -0,0 +1,35 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2024, STMicroelectronics
+ */
+
+#ifndef SCMI_SERVER_CLOCK_CONSUMER_H
+#define SCMI_SERVER_CLOCK_CONSUMER_H
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <tee_api_types.h>
+#include <scmi_agent_configuration.h>
+
+#ifdef CFG_SCMI_SERVER_CLOCK_CONSUMER
+/*
+ * Initialize SCMI clock from clock consumer information from DT.
+ * Returns a TEE_Result compliant value
+ */
+TEE_Result optee_scmi_server_init_clocks(const void *fdt, int node,
+					 struct scpfw_agent_config *agent_cfg,
+					 struct scpfw_channel_config
+						*channel_cfg);
+#else
+static inline TEE_Result
+	optee_scmi_server_init_clocks(const void *fdt __unused,
+				      int node __unused,
+				      struct scpfw_agent_config *agent_cfg
+						__unused,
+				      struct scpfw_channel_config *channel_cfg
+						__unused)
+{
+	return TEE_SUCCESS;
+}
+#endif /* CFG_SCMI_SERVER_CLOCK_CONSUMER */
+#endif /* SCMI_SERVER_CLOCK_CONSUMER_H */
diff --git a/core/lib/scmi-server/scmi_clock_consumer.c b/core/lib/scmi-server/scmi_clock_consumer.c
new file mode 100644
index 000000000..32450cd66
--- /dev/null
+++ b/core/lib/scmi-server/scmi_clock_consumer.c
@@ -0,0 +1,278 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2024, STMicroelectronics
+ */
+
+#include <assert.h>
+#include <drivers/clk.h>
+#include <drivers/clk_dt.h>
+#include <kernel/boot.h>
+#include <kernel/panic.h>
+#include <libfdt.h>
+#include <malloc.h>
+#include <scmi_agent_configuration.h>
+#include <scmi_clock_consumer.h>
+#include <tee_api_defines_extensions.h>
+#include <dt-bindings/scmi/scmi-clock.h>
+#include <trace.h>
+
+/*
+ * struct scmi_server_clock: data for a SCMI clock in DT
+ *
+ * @domain_id: SCMI domain identifier
+ * @domain_name: SCMI domain name
+ * @clock: clock to control thru SCMI protocol
+ * @flags: capabilities of the SCMI clock
+ */
+struct scmi_server_clock {
+	uint32_t domain_id;
+	const char *domain_name;
+	struct clk *clock;
+	uint32_t flags;
+};
+
+/*
+ * struct scmi_clock_consumer: context of scmi_clock_consumer.c for one channel
+ *
+ * @plat_clocks: platform clocks exposed with SCMI
+ * @plat_clocks_flags: capabilities of an plat_clocks
+ * @plat_clocks_count: number of plat_clocks and plat_clocks_flags
+ * @link: link for list of scmi_clock_consumer
+ */
+struct scmi_clock_consumer {
+	struct clk *plat_clocks;
+	uint32_t *plat_clocks_flags;
+	unsigned int plat_clocks_count;
+
+	SLIST_ENTRY(scmi_clock_consumer) link;
+};
+
+static SLIST_HEAD(scmi_clock_consumer_head, scmi_clock_consumer) ctx =
+	SLIST_HEAD_INITIALIZER(&ctx);
+
+static TEE_Result plat_scmi_clk_get_rates_array(struct clk *clk, size_t index,
+						unsigned long *rates,
+						size_t *nb_elts)
+{
+	uint32_t plat_clock_flags = *(uint32_t *)clk->priv;
+
+	if (!nb_elts)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (plat_clock_flags & SCMI_CLOCK_ALLOW_SET_RATE)
+		return clk_get_rates_array(clk->parent, index, rates, nb_elts);
+
+	if (!rates || !*nb_elts) {
+		*nb_elts = 1;
+		return TEE_SUCCESS;
+	}
+
+	if (index)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	assert(rates);
+
+	/*
+	 * Clocks not exposed have no effective parent/platform clock.
+	 * Report a 0 Hz rate in this case.
+	 */
+	if (clk->parent)
+		*rates = clk_get_rate(clk->parent);
+	else
+		*rates = 0;
+
+	*nb_elts = 1;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result plat_scmi_clk_get_rates_steps(struct clk *clk,
+						unsigned long *min,
+						unsigned long *max,
+						unsigned long *step)
+{
+	uint32_t plat_clock_flags = *(uint32_t *)clk->priv;
+
+	if (plat_clock_flags & SCMI_CLOCK_ALLOW_SET_RATE) {
+		*min = 0;
+		*max = UINT32_MAX;
+		*step = 0;
+	} else {
+		*min = clk->rate;
+		*max = *min;
+		*step = 0;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static const struct clk_ops plat_scmi_clk_ops = {
+	.get_rates_array = plat_scmi_clk_get_rates_array,
+	.get_rates_steps = plat_scmi_clk_get_rates_steps,
+};
+
+TEE_Result optee_scmi_server_init_clocks(const void *fdt, int node,
+					 struct scpfw_agent_config *agent_cfg,
+					 struct scpfw_channel_config
+						*channel_cfg)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct scmi_clock_consumer *consumer = NULL;
+	struct scmi_server_clock *s_clocks = NULL;
+	size_t s_clocks_count = 0;
+	int subnode = 0, item_node = 0;
+	bool have_subnodes = false;
+	size_t n = 0;
+	size_t i = 0;
+
+	item_node = fdt_subnode_offset(fdt, node, "clocks");
+	if (item_node < 0)
+		return TEE_SUCCESS;
+
+	/* Compute the number of domains to allocate */
+	fdt_for_each_subnode(subnode, fdt, item_node) {
+		paddr_t reg = fdt_reg_base_address(fdt, subnode);
+
+		assert(reg != DT_INFO_INVALID_REG);
+		if ((size_t)reg > s_clocks_count)
+			s_clocks_count = (uint32_t)reg;
+
+		have_subnodes = true;
+	}
+
+	if (!have_subnodes)
+		return TEE_SUCCESS;
+
+	consumer = calloc(1, sizeof(*consumer));
+	if (!consumer)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	SLIST_INSERT_HEAD(&ctx, consumer, link);
+
+	s_clocks_count++;
+	s_clocks = calloc(s_clocks_count, sizeof(*s_clocks));
+	if (!s_clocks) {
+		SLIST_REMOVE_HEAD(&ctx, link);
+		free(consumer);
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	fdt_for_each_subnode(subnode, fdt, item_node) {
+		struct scmi_server_clock *s_clock = NULL;
+		const fdt32_t *cuint = NULL;
+		struct clk *clock = NULL;
+		uint32_t domain_id = 0;
+
+		res = clk_dt_get_by_index(fdt, subnode, 0, &clock);
+		if (res == TEE_ERROR_DEFER_DRIVER_INIT) {
+			panic("Unexpected init deferral");
+		} else if (res) {
+			EMSG("Can't get clock %s (%#"PRIx32"), skipped",
+			     fdt_get_name(fdt, subnode, NULL), res);
+			continue;
+		}
+
+		domain_id = (uint32_t)fdt_reg_base_address(fdt, subnode);
+		s_clock = s_clocks + domain_id;
+		s_clock->domain_id = domain_id;
+
+		cuint = fdt_getprop(fdt, subnode, "domain-name", NULL);
+		if (cuint)
+			s_clock->domain_name = (char *)cuint;
+		else
+			s_clock->domain_name = fdt_get_name(fdt, subnode, NULL);
+
+		s_clock->domain_name = strdup(s_clock->domain_name);
+		if (!s_clock->domain_name) {
+			free(s_clocks);
+			SLIST_REMOVE_HEAD(&ctx, link);
+			free(consumer);
+			return TEE_ERROR_OUT_OF_MEMORY;
+		}
+
+		/* Check that the domain_id is not already used */
+		if (s_clock->clock) {
+			EMSG("Domain ID %"PRIu32" already used", domain_id);
+			panic();
+		}
+		s_clock->clock = clock;
+
+		/*
+		 * Get clock flags
+		 */
+		cuint = fdt_getprop(fdt, subnode, "flags", NULL);
+		if (cuint) {
+			s_clock->flags = fdt32_to_cpu(*cuint);
+			s_clock->flags &= SCMI_CLOCK_MASK;
+		}
+
+		DMSG("scmi clock shares %s on domain ID %"PRIu32,
+		     s_clock->domain_name, domain_id);
+	}
+
+	/*
+	 * Assign domain IDs to un-exposed clock as SCMI specification
+	 * require the resource is defined even if not accessible.
+	 */
+	for (n = 0; n < s_clocks_count; n++) {
+		if (!s_clocks[n].clock) {
+			s_clocks[n].domain_id = n;
+			s_clocks[n].domain_name = "";
+		}
+	}
+
+	if (consumer->plat_clocks || channel_cfg->clock) {
+		EMSG("Clock already loaded: agent %u, channel %u",
+		     agent_cfg->agent_id, channel_cfg->channel_id);
+		panic();
+	}
+
+	consumer->plat_clocks_count = s_clocks_count;
+	consumer->plat_clocks = calloc(consumer->plat_clocks_count,
+				       sizeof(*consumer->plat_clocks));
+	assert(consumer->plat_clocks);
+
+	consumer->plat_clocks_flags =
+		calloc(consumer->plat_clocks_count,
+		       sizeof(*consumer->plat_clocks_flags));
+	assert(consumer->plat_clocks_flags);
+
+	channel_cfg->clock_count = s_clocks_count;
+	channel_cfg->clock = calloc(channel_cfg->clock_count,
+				    sizeof(*channel_cfg->clock));
+	assert(channel_cfg->clock);
+
+	for (i = 0; i < s_clocks_count; i++) {
+		unsigned int domain_id = s_clocks[i].domain_id;
+		struct clk *plat_clock = consumer->plat_clocks + domain_id;
+		uint32_t *plat_clock_flags = consumer->plat_clocks_flags +
+			domain_id;
+
+		*plat_clock_flags = s_clocks[i].flags;
+
+		plat_clock->ops = &plat_scmi_clk_ops;
+		plat_clock->priv = plat_clock_flags;
+		plat_clock->name = s_clocks[i].domain_name;
+		plat_clock->parent = s_clocks[i].clock;
+
+		if (*plat_clock_flags & SCMI_CLOCK_ALLOW_SET_RATE)
+			plat_clock->flags = CLK_SET_RATE_PARENT;
+
+		plat_clock->flags |= CLK_DUTY_CYCLE_PARENT;
+
+		if (clk_register(plat_clock))
+			panic();
+
+		channel_cfg->clock[domain_id] = (struct scmi_clock){
+			.name = clk_get_name(plat_clock),
+			.clk = plat_clock,
+			/* Default OFF for RIF issues */
+			.enabled = false,
+		};
+	}
+
+	/* We can free s_clk_channel resources since SCMI server handles them */
+	free(s_clocks);
+
+	return TEE_SUCCESS;
+}
diff --git a/core/lib/scmi-server/sub.mk b/core/lib/scmi-server/sub.mk
index 13b9974c2..bfbc8bc0e 100644
--- a/core/lib/scmi-server/sub.mk
+++ b/core/lib/scmi-server/sub.mk
@@ -26,6 +26,7 @@ scpfw-integ-version-pat = 0
 scpfw-integ-version = $(scpfw-integ-version-maj).$(scpfw-integ-version-min).$(scpfw-integ-version-pat)
 
 srcs-y += scmi_server.c
+srcs-$(CFG_SCMI_SERVER_CLOCK_CONSUMER) += scmi_clock_consumer.c
 srcs-$(CFG_SCMI_SERVER_PD_CONSUMER) += scmi_pd_consumer.c
 srcs-$(CFG_SCMI_SERVER_REGULATOR_CONSUMER) += scmi_regulator_consumer.c
 global-incdirs-y += include
-- 
2.47.2

